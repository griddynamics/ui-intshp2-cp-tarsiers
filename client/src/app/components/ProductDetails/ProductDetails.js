import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Spinner from '../../shared/Spinner';
import { productType } from '../../types';
import {
  isAddedToCart,
  isAddedToWishList
} from '../../../utils/inCartInWishlist.service';
import ProductDescriptionContainer from './ProductDescription/ProductDescriptionContainer';
import RelatedProducts from './RelatedProducts/RelatedProducts';
import ImagePreview from '../ImagePreview/ImagePreview';
import HttpService from '../../../utils/http.service';
import appConfig from '../../../config/appConfig';

const { products } = appConfig.apiResources;

class ProductDetails extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    wishlist: PropTypes.arrayOf(productType),
    cart: PropTypes.shape({
      productsInCart: PropTypes.arrayOf(productType)
    })
  };

  static defaultProps = { wishlist: [], cart: { productsInCart: [] } };

  state = { item: null };

  componentDidMount = () => {
    window.scrollTo(0, 0);
    this.updateMainProduct();
  };

  componentDidUpdate = () => {
    const { item } = this.state;
    const { id } = this.props;

    if (item) {
      if (item._id !== id) {
        this.updateMainProduct();
      }
    }
  };

  updateMainProduct = () => {
    const { id } = this.props;

    const params = {
      _id: id
    };

    HttpService.get(products, { params }).then(res => {
      this.setState({ item: res.data[0] });
    });
  };

  render() {
    const { id, wishlist, cart } = this.props;
    const { item } = this.state;
    const wished = isAddedToWishList(id, wishlist);
    const inCart = isAddedToCart(id, cart);

    return item ? (
      <React.Fragment>
        <ProductDescriptionContainer
          item={item}
          wished={wished}
          inCart={inCart}
        >
          <ImagePreview item={item} />
        </ProductDescriptionContainer>
        <RelatedProducts item={item} wishlist={wishlist} cart={cart} />
      </React.Fragment>
    ) : (
      <div className="container">
        <Spinner height="80vh" />
      </div>
    );
  }
}

export default ProductDetails;
