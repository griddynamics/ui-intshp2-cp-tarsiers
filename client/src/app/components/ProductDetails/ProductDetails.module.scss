@import '../../../styles/mixin';
@import '../../../styles/variables';

.section {
  @include d-flex(center, row);
  background-color: $color-white;
  border-top: 0.5px solid $color-light-2;
  padding-bottom: 75px;
}

.innersection {
  @include d-flex(center, row);
  max-width: $media-large;
}

@media screen and (min-width: $media-tablet) and (max-width: $media-medium) {
  .innersection {
    max-width: $media-tablet;
  }
}

@media screen and (min-width: $media-smallest) and (max-width: $media-tablet) {
  .innersection {
    display: block;
  }
}
