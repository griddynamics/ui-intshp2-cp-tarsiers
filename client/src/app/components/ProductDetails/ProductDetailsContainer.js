import { connect } from 'react-redux';
import ProductDetails from './ProductDetails';

const mapStateToProps = state => ({
  products: state.products.products,
  wishlist: state.wishlist.wishlist,
  cart: state.cart
});
const ProductDetailsContainer = connect(mapStateToProps)(ProductDetails);

export default ProductDetailsContainer;
