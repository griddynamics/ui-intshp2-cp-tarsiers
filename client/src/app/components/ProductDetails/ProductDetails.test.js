import React from 'react';
import { shallow } from 'enzyme';
import ProductDetails from './ProductDetails';

const products = require('../../../mocks/products.json');

describe('<ProductDetails />', () => {
  let props;

  beforeEach(() => {
    props = {
      id: '5c47312be7179a544940e080',
      products,
      cart: {
        value: 0,
        productsInCart: []
      }
    };
  });

  it('should match its snapshot', () => {
    const wrapper = shallow(<ProductDetails {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
