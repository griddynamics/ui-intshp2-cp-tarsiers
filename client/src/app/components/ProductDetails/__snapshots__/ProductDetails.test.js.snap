// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> should match its snapshot 1`] = `
<div
  className="container"
>
  <Spinner
    height="80vh"
  />
</div>
`;
